src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -0,0 +1,1211 @@
+//include header file
+#include "hiopAlgPrimalDecomp.hpp"
+#include "hiopInterfacePrimalDecomp.hpp"
+
+#include <cstring>
+#include <cassert>
+
+namespace hiop
+{
+#ifdef HIOP_USE_MPI
+  /* This struct provides the info necessary for the recourse approximation function
+   * buffer[n+1] contains both the function value and gradient w.r.t x
+   * buffer[0] is the function value and buffer[1:n] the gradient
+   * Contains send and receive functionalities for the values in buffer
+   */
+  struct ReqRecourseApprox
```

Comment:
Consider refactoring this class to contain wrappers for MPI calls and a pointer to buffer packing/unpacking object. Buffer management class should have an abstract interface and should be created using a factory. For this PR it would suffice to have just one buffer manager implementation.

Reply:
Also, probably good idea would be to put `ReqRecourseApprox` class implementation in a separate source file.

Reply:
Hoping to address this in discussion this week.

Reply:
Per discussion, use `hiopVector` as a container for buffer data.